plugins/http_plugin/beast_http_session.hpp

```diff
@@ -0,0 +1,587 @@
+                send_response(std::string(what), 
+                            static_cast<int>(http::status::too_many_requests));                
+            }
+
+        public:
+            virtual bool verify_max_bytes_in_flight() override {
+                auto bytes_in_flight_size = plugin_state_->bytes_in_flight.load();
+                if( bytes_in_flight_size > plugin_state_->max_bytes_in_flight ) {
+                    fc_dlog( logger, "429 - too many bytes in flight: ${bytes}", ("bytes", bytes_in_flight_size) );
+                    string what = "Too many bytes in flight: " + std::to_string( bytes_in_flight_size ) + ". Try again later.";;
+                    report_429_error(what);
+                    return false;
+                }
+                return true;
+            }
+
+            virtual bool verify_max_requests_in_flight() override {
+                if (plugin_state_->max_requests_in_flight < 0)
+                    return true;
+
+                auto requests_in_flight_num = plugin_state_->requests_in_flight.load();
+                if( requests_in_flight_num > plugin_state_->max_requests_in_flight ) {
+                    fc_dlog( logger, "429 - too many requests in flight: ${requests}", ("requests", requests_in_flight_num) );
+                    string what = "Too many requests in flight: " + std::to_string( requests_in_flight_num ) + ". Try again later.";
+                    report_429_error(what);
+                    return false;
+                }
+                return true;
+            }           
+
+            // Access the derived class, this is part of
+            // the Curiously Recurring Template Pattern idiom.
+            Derived& derived()
+            {
+                return static_cast<Derived&>(*this);
+            }
+            
+        public: 
+            // shared_from_this() requires default constuctor
+            beast_http_session() = default;
+      
+            // Take ownership of the buffer
+            beast_http_session(
+                std::shared_ptr<http_plugin_state> plugin_state,  
+                std::shared_ptr<eosio::chain::named_thread_pool> thread_pool)
+                : plugin_state_(plugin_state) 
+                , thread_pool_(thread_pool)
+            {  
+                plugin_state_->requests_in_flight += 1;
+                req_parser_.emplace();
+                req_parser_->body_limit(plugin_state_->max_body_size);
+
+                session_begin_ = steady_clock::now();
+                read_time_us_ = handle_time_us_ = write_time_us_ = 0;
+
+                // dfeault to true
+                isSendExceptionResponse_ = true;
+            }
+
+            virtual ~beast_http_session() {
+                // avoid exceptions being thrown in destructor
+                isSendExceptionResponse_ = false;
+                try { 
+                    plugin_state_->requests_in_flight -= 1;
+    #if PRINT_PERF_METRICS                
```

Comment:
I don't think this should be a `#if`. It could be left as debug level logging always. Or if you think debug is too much you can do:
     ` if( logger.is_enabled( fc::log_level::all ) )`